src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListenerBase.cs

```diff
@@ -272,22 +286,123 @@ private static async Task ProcessWorkCoreAsync(ImmutableArray<Work> work, Stream
+            return;
+        }
+
+        if (!_projectEntryMap.TryGetValue(project.Id, out var entry))
+        {
+            _projectEntryMap[project.Id] = entry = new();
+        }
+
+        var (didChange, projectInfo) = await DidProjectInfoChangeAsync(project, entry, projectPath, intermediateOutputPath, cancellationToken).ConfigureAwait(false);
+        if (didChange)
+        {
+            projectInfo.AssumeNotNull();
+            UpdateEntry(project, projectInfo, entry);
+
+            stream.WriteProjectInfoAction(ProjectInfoAction.Update);
+            await stream.WriteProjectInfoAsync(projectInfo, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    private void UpdateEntry(Project project, RazorProjectInfo projectInfo, ProjectEntry entry)
+    {
+        entry.RootNamespace = projectInfo.RootNamespace;
+        entry.ConfigurationHash = projectInfo.Configuration.GetHashCode();
+        entry.DocumentsHash = projectInfo.Documents.GetHashCode();
+
+        _cachedTagHelperResolver.TryGetId(project.Id, out var resultId);
+        entry.TagHelpersResultId = resultId;
+    }
+
+    private async Task<(bool, RazorProjectInfo? projectInfo)> DidProjectInfoChangeAsync(Project project, ProjectEntry entry, string projectPath, string intermediateOutputPath, CancellationToken cancellationToken)
+    {
+        ImmutableArray<DocumentSnapshotHandle> documents = default;
+        var csharpLanguageVersion = (project.ParseOptions as CSharpParseOptions)?.LanguageVersion ?? LanguageVersion.Default;
+        CachedTagHelperResolver.DeltaResult? delta = null;
+
+        var configuration = RazorProjectInfoHelpers.ComputeRazorConfigurationOptions(project.AnalyzerOptions.AnalyzerConfigOptionsProvider, out var rootNamespace);
+        if (configuration is null)
+        {
+            return (false, null);
+        }
+
+        if (entry.RootNamespace is null)
+        {
+            return await GetChangedValueAsync().ConfigureAwait(false);
+        }
+
+        if (!entry.RootNamespace.Equals(rootNamespace))
+        {
+            return await GetChangedValueAsync().ConfigureAwait(false);
+        }
+
+        if (entry.ConfigurationHash != configuration.GetHashCode())
+        {
+            return await GetChangedValueAsync().ConfigureAwait(false);
+        }
+
+        documents = RazorProjectInfoHelpers.GetDocuments(project, projectPath);
+        if (documents.Length == 0)
+        {
+            _logger.LogInformation("No razor documents in {projectId}", project.Id);
+            return (false, null);
+        }
+
+        if (entry.DocumentsHash != documents.GetHashCode())
+        {
+            return await GetChangedValueAsync().ConfigureAwait(false);
+        }
+
+        delta = await _cachedTagHelperResolver.GetDeltaAsync(project, entry.TagHelpersResultId, cancellationToken).ConfigureAwait(false);
+        if (delta.IsDelta)
+        {
+            return await GetChangedValueAsync().ConfigureAwait(false);
```

Comment:
This is very much personal opinion, but there is something about the repeated calls to `GetChangedValueAsync` that confuses me at first glance. Consider flipping all of these `if`s so they're testing for the unchanged state, and returning false, and leave the computation of the new value to the end.

Reply:
I like it, changing